File: desktop/src/features/cos-user-context/lib/cosUserContext.test.mjs
import assert from "node:assert/strict";
import test from "node:test";
import { finalizeEvent, getPublicKey } from "nostr-tools/pure";

import {
  currentCosUserContext,
  cosUserContextChannelCandidates,
  hasCosWorkspaceModule,
  parseCosUserContext,
  resolveAuthoritativeCosUserContextChannel,
  selectLatestCosUserContext,
} from "./cosUserContext.ts";

const assignee = "a".repeat(64);
const bridgeSecret = new Uint8Array(32).fill(1);
const attackerSecret = new Uint8Array(32).fill(2);
const relaySecret = new Uint8Array(32).fill(3);
const bridge = getPublicKey(bridgeSecret);
const relay = getPublicKey(relaySecret);
const channel = "550e8400-e29b-41d4-a716-446655440000";

function contextEvent({
  secretKey = bridgeSecret,
  createdAt = 1,
  modules = ["today", "my_actions", "messages", "assistant"],
  channelId = channel,
} = {}) {
  return finalizeEvent(
    {
      created_at: createdAt,
      kind: 37012,
      tags: [
        ["h", channelId],
        ["d", `context:${assignee}`],
        ["p", assignee],
      ],
      content: JSON.stringify({
        schema: "mac-workspace/cos-user-context/v1",
        tenant_slug: "mac-surfacing",
        user: {
          id: 42,
          name: "Jake Wherton",
          role: "contractor_admin",
          role_label: "Leadership",
        },
        modules,
        assistant: modules.includes("assistant")
          ? {
              key: "mac-assistant",
              label: "MAC Assistant",
              execution: "brain-vps",
              memory_scope: "private-channel",
            }
          : null,
        generated_at: "2026-07-28T08:00:00Z",
      }),
    },
    secretKey,
  );
}

function channelMetadata(channelId, { isPrivate = true, createdAt = 1 } = {}) {
  return finalizeEvent(
    {
      created_at: createdAt,
      kind: 39000,
      tags: [["d", channelId], ...(isPrivate ? [["private"]] : [["public"]])],
      content: "",
    },
    relaySecret,
  );
}

function channelMembership(
  channelId,
  {
    assigneePubkey = assignee,
    ownerPubkey = bridge,
    extraPubkey,
    createdAt = 1,
  } = {},
) {
  return finalizeEvent(
    {
      created_at: createdAt,
      kind: 39002,
      tags: [
        ["d", channelId],
        ["p", ownerPubkey, "", "owner"],
        ["p", assigneePubkey, "", "member"],
        ...(extraPubkey ? [["p", extraPubkey, "", "member"]] : []),
      ],
      content: "",
    },
    relaySecret,
  );
}

test("parses a private staff context and exposes only projected modules", () => {
  const context = parseCosUserContext(contextEvent(), assignee);
  assert.equal(context.user.name, "Jake Wherton");
  assert.equal(hasCosWorkspaceModule(context, "assistant"), true);
  assert.equal(hasCosWorkspaceModule(context, "agents"), false);
});

test("accepts least-privilege module subsets and rejects malformed modules", () => {
  const restricted = parseCosUserContext(
    contextEvent({ modules: ["today", "messages"] }),
    assignee,
  );
  assert.equal(hasCosWorkspaceModule(restricted, "my_actions"), false);
  assert.throws(() =>
    parseCosUserContext(
      contextEvent({ modules: ["today", "secrets"] }),
      assignee,
    ),
  );
  assert.throws(() =>
    parseCosUserContext(
      {
        ...contextEvent(),
        tags: [
          ["h", channel],
          ["d", `context:${"d".repeat(64)}`],
          ["p", "d".repeat(64)],
        ],
      },
      assignee,
    ),
  );
  assert.throws(
    () =>
      parseCosUserContext(
        { ...contextEvent(), content: '{"schema":"tampered"}' },
        assignee,
      ),
    /signature is invalid/,
  );
});

test("selects the latest projection from only the trusted bridge", () => {
  const first = contextEvent();
  const second = contextEvent({
    createdAt: 2,
    modules: [
      "today",
      "my_actions",
      "messages",
      "assistant",
      "running_order",
      "agents",
    ],
  });
  const latest = selectLatestCosUserContext(
    [
      first,
      second,
      contextEvent({
        secretKey: attackerSecret,
        createdAt: 3,
        modules: [
          "today",
          "my_actions",
          "messages",
          "assistant",
          "running_order",
          "agents",
        ],
      }),
    ],
    assignee,
    bridge,
  );
  assert.equal(latest?.eventId, second.id);
});

test("a newer trusted module revocation replaces the older projection", () => {
  const older = contextEvent({ createdAt: 1 });
  const restricted = contextEvent({
    createdAt: 2,
    modules: ["today", "messages"],
  });
  const latest = selectLatestCosUserContext(
    [older, restricted],
    assignee,
    bridge,
  );
  assert.equal(latest?.eventId, restricted.id);
  assert.equal(hasCosWorkspaceModule(latest, "my_actions"), false);
});

test("binds role projection to the unique current private identity channel", () => {
  const otherChannel = "550e8400-e29b-41d4-a716-446655440001";
  const events = [
    contextEvent({ channelId: channel }),
    contextEvent({ channelId: otherChannel, createdAt: 2 }),
  ];
  const candidates = cosUserContextChannelCandidates(events, assignee, bridge);
  assert.deepEqual(candidates, [channel, otherChannel]);
  const resolved = resolveAuthoritativeCosUserContextChannel({
    candidateChannelIds: candidates,
    metadataEvents: [channelMetadata(channel), channelMetadata(otherChannel)],
    membershipEvents: [
      channelMembership(channel),
      channelMembership(otherChannel, { extraPubkey: "c".repeat(64) }),
    ],
    assigneePubkey: assignee,
    trustedBridgePubkey: bridge,
    trustedRelayPubkey: relay,
  });
  assert.equal(resolved, channel);
  assert.equal(
    selectLatestCosUserContext(events, assignee, bridge, resolved)?.channelId,
    channel,
  );
});

test("fails closed when two channels claim the same active identity", () => {
  const otherChannel = "550e8400-e29b-41d4-a716-446655440001";
  const candidates = [channel, otherChannel];
  assert.equal(
    resolveAuthoritativeCosUserContextChannel({
      candidateChannelIds: candidates,
      metadataEvents: [channelMetadata(channel), channelMetadata(otherChannel)],
      membershipEvents: [
        channelMembership(channel),
        channelMembership(otherChannel),
      ],
      assigneePubkey: assignee,
      trustedBridgePubkey: bridge,
      trustedRelayPubkey: relay,
    }),
    null,
  );
  assert.throws(
    () =>
      parseCosUserContext(
        contextEvent({ channelId: otherChannel }),
        assignee,
        channel,
      ),
    /different channel/,
  );
});

test("rejects channel authority not signed by the active relay", () => {
  const forgedMetadata = finalizeEvent(
    {
      created_at: 2,
      kind: 39000,
      tags: [["d", channel], ["private"]],
      content: "",
    },
    attackerSecret,
  );
  assert.equal(
    resolveAuthoritativeCosUserContextChannel({
      candidateChannelIds: [channel],
      metadataEvents: [forgedMetadata],
      membershipEvents: [channelMembership(channel)],
      assigneePubkey: assignee,
      trustedBridgePubkey: bridge,
      trustedRelayPubkey: relay,
    }),
    null,
  );
});

test("does not authorise cached context after a refresh failure", () => {
  const cached = parseCosUserContext(contextEvent(), assignee);
  assert.equal(
    currentCosUserContext({
      data: cached,
      isError: false,
      isFetching: false,
      isPending: false,
    }),
    cached,
  );
  assert.equal(
    currentCosUserContext({
      data: cached,
      isError: true,
      isFetching: false,
      isPending: false,
    }),
    null,
  );
  assert.equal(
    currentCosUserContext({
      data: cached,
      isError: false,
      isFetching: false,
      isPending: true,
    }),
    null,
  );
  assert.equal(
    currentCosUserContext({
      data: cached,
      isError: false,
      isFetching: true,
      isPending: false,
    }),
    null,
  );
});

File: mobile/test/features/cos_user_context/cos_user_context_test.dart
import 'dart:convert';

import 'package:buzz/features/cos_user_context/cos_user_context.dart';
import 'package:buzz/shared/relay/nostr_models.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:nostr/nostr.dart' as nostr;

void main() {
  const assignee =
      'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa';
  const bridgeSecret =
      '0101010101010101010101010101010101010101010101010101010101010101';
  const attackerSecret =
      '0202020202020202020202020202020202020202020202020202020202020202';
  const relaySecret =
      '0303030303030303030303030303030303030303030303030303030303030303';
  final bridge = nostr.Keys(bridgeSecret).public;
  final relay = nostr.Keys(relaySecret).public;

  NostrEvent event({
    int createdAt = 1,
    String secretKey = bridgeSecret,
    String channelId = 'mac',
    List<String> modules = const [
      'today',
      'my_actions',
      'messages',
      'agents',
      'running_order',
    ],
    Map<String, Object?>? content,
  }) {
    final signed = nostr.Event.from(
      kind: EventKind.cosUserContext,
      content: jsonEncode(
        content ??
            {
              'schema': 'mac-workspace/cos-user-context/v1',
              'tenant_slug': 'mac-surfacing',
              'user': {
                'id': 42,
                'name': 'Jake Wherton',
                'role': 'managing_director',
                'role_label': 'Managing Director',
              },
              'modules': modules,
              'assistant': modules.contains('assistant')
                  ? {
                      'key': 'mac-assistant',
                      'label': 'MAC Assistant',
                      'execution': 'brain-vps',
                      'memory_scope': 'private-channel',
                    }
                  : null,
              'generated_at': '2026-07-29T09:00:00Z',
            },
      ),
      tags: [
        ['h', channelId],
        const ['p', assignee],
        const ['d', 'context:$assignee'],
      ],
      secretKey: secretKey,
      createdAt: createdAt,
      verify: true,
    );
    return NostrEvent.fromJson(signed.toMap());
  }

  NostrEvent channelState({
    required int kind,
    required String channelId,
    required List<List<String>> tags,
    String secretKey = relaySecret,
    int createdAt = 1,
  }) {
    final signed = nostr.Event.from(
      kind: kind,
      content: '',
      tags: [
        ['d', channelId],
        ...tags,
      ],
      secretKey: secretKey,
      createdAt: createdAt,
      verify: true,
    );
    return NostrEvent.fromJson(signed.toMap());
  }

  test('requires both technical modules for Control Room', () {
    final context = parseCosUserContext(event(), expectedAssignee: assignee);
    expect(context.canUseControlRoom, isTrue);
  });

  test('accepts restricted modules and fails closed for untrusted context', () {
    expect(
      selectLatestCosUserContext(
        [event(secretKey: attackerSecret)],
        assigneePubkey: assignee,
        trustedBridgePubkey: bridge,
      ),
      isNull,
    );
    expect(
      parseCosUserContext(
        event(modules: const ['today', 'messages']),
        expectedAssignee: assignee,
      ).hasModule('my_actions'),
      isFalse,
    );
    final signed = event();
    expect(
      () => parseCosUserContext(
        NostrEvent(
          id: signed.id,
          pubkey: signed.pubkey,
          createdAt: signed.createdAt,
          kind: signed.kind,
          tags: signed.tags,
          content: '${signed.content} ',
          sig: signed.sig,
        ),
        expectedAssignee: assignee,
      ),
      throwsFormatException,
    );
  });

  test('fails closed when canonical identity fields are missing', () {
    for (final content in [
      {
        'schema': 'mac-workspace/cos-user-context/v1',
        'modules': ['today', 'my_actions', 'messages'],
      },
      {
        'schema': 'mac-workspace/cos-user-context/v1',
        'tenant_slug': 'MAC Surfacing',
        'user': {
          'id': 42,
          'name': 'Jake Wherton',
          'role': 'managing_director',
          'role_label': 'Managing Director',
        },
        'modules': ['today', 'my_actions', 'messages'],
        'assistant': null,
        'generated_at': '2026-07-29T09:00:00Z',
      },
      {
        'schema': 'mac-workspace/cos-user-context/v1',
        'tenant_slug': 'mac--surfacing',
        'user': {
          'id': 42,
          'name': 'Jake Wherton',
          'role': 'managing_director',
          'role_label': 'Managing Director',
        },
        'modules': ['today', 'messages'],
        'assistant': null,
        'generated_at': '2026-07-29T09:00:00Z',
      },
      {
        'schema': 'mac-workspace/cos-user-context/v1',
        'tenant_slug': 'mac-surfacing',
        'user': {
          'id': null,
          'name': 'Jake Wherton',
          'role': 'managing_director',
          'role_label': 'Managing Director',
        },
        'modules': ['today', 'my_actions', 'messages'],
        'assistant': null,
        'generated_at': '2026-07-29T09:00:00Z',
      },
      {
        'schema': 'mac-workspace/cos-user-context/v1',
        'tenant_slug': 'mac-surfacing',
        'user': {
          'id': 42,
          'name': 'Jake Wherton',
          'role': 'managing_director',
          'role_label': 'Managing Director',
        },
        'modules': ['today', 'my_actions', 'messages'],
        'assistant': null,
        'generated_at': ' ',
      },
    ]) {
      expect(
        () => parseCosUserContext(
          event(content: content),
          expectedAssignee: assignee,
        ),
        throwsFormatException,
      );
    }
  });

  test('enforces the private assistant boundary', () {
    final invalidAssistant = {
      'schema': 'mac-workspace/cos-user-context/v1',
      'tenant_slug': 'mac-surfacing',
      'user': {
        'id': 42,
        'name': 'Jake Wherton',
        'role': 'managing_director',
        'role_label': 'Managing Director',
      },
      'modules': ['today', 'my_actions', 'messages', 'assistant'],
      'assistant': {
        'key': 'mac-assistant',
        'label': 'MAC Assistant',
        'execution': 'on-device',
        'memory_scope': 'shared',
      },
      'generated_at': '2026-07-29T09:00:00Z',
    };
    expect(
      () => parseCosUserContext(
        event(content: invalidAssistant),
        expectedAssignee: assignee,
      ),
      throwsFormatException,
    );
    expect(
      parseCosUserContext(
        event(modules: const ['today', 'my_actions', 'messages', 'assistant']),
        expectedAssignee: assignee,
      ).hasModule('assistant'),
      isTrue,
    );
  });

  test('selects the newest trusted projection deterministically', () {
    final older = event(createdAt: 1);
    final newer = event(createdAt: 2, modules: const ['today', 'messages']);
    final selected = selectLatestCosUserContext(
      [older, newer],
      assigneePubkey: assignee,
      trustedBridgePubkey: bridge,
    );
    expect(selected?.eventId, newer.id);
    expect(selected?.hasModule('my_actions'), isFalse);
  });

  test('binds access to one exact private bridge-owned channel', () {
    const otherChannel = 'other';
    final contexts = [
      event(channelId: 'mac'),
      event(channelId: otherChannel, createdAt: 2),
    ];
    final candidates = cosUserContextChannelCandidates(
      contexts,
      assigneePubkey: assignee,
      trustedBridgePubkey: bridge,
    );
    expect(candidates, ['mac', otherChannel]);
    final resolved = resolveAuthoritativeCosUserContextChannel(
      candidateChannelIds: candidates,
      metadataEvents: [
        channelState(
          kind: 39000,
          channelId: 'mac',
          tags: const [
            ['private'],
          ],
        ),
        channelState(
          kind: 39000,
          channelId: otherChannel,
          tags: const [
            ['private'],
          ],
        ),
      ],
      membershipEvents: [
        channelState(
          kind: 39002,
          channelId: 'mac',
          tags: [
            ['p', bridge, '', 'owner'],
            const ['p', assignee, '', 'member'],
          ],
        ),
        channelState(
          kind: 39002,
          channelId: otherChannel,
          tags: [
            ['p', bridge, '', 'owner'],
            const ['p', assignee, '', 'member'],
            const ['p', attackerSecret, '', 'member'],
          ],
        ),
      ],
      assigneePubkey: assignee,
      trustedBridgePubkey: bridge,
      trustedRelayPubkey: relay,
    );
    expect(resolved, 'mac');
    expect(
      selectLatestCosUserContext(
        contexts,
        assigneePubkey: assignee,
        trustedBridgePubkey: bridge,
        expectedChannelId: resolved,
      )?.channelId,
      'mac',
    );
  });

  test('fails closed when two channels claim the same identity', () {
    const otherChannel = 'other';
    final resolved = resolveAuthoritativeCosUserContextChannel(
      candidateChannelIds: const ['mac', otherChannel],
      metadataEvents: [
        channelState(
          kind: 39000,
          channelId: 'mac',
          tags: const [
            ['private'],
          ],
        ),
        channelState(
          kind: 39000,
          channelId: otherChannel,
          tags: const [
            ['private'],
          ],
        ),
      ],
      membershipEvents: [
        for (final channel in const ['mac', otherChannel])
          channelState(
            kind: 39002,
            channelId: channel,
            tags: [
              ['p', bridge, '', 'owner'],
              const ['p', assignee, '', 'member'],
            ],
          ),
      ],
      assigneePubkey: assignee,
      trustedBridgePubkey: bridge,
      trustedRelayPubkey: relay,
    );
    expect(resolved, isNull);
    expect(
      () => parseCosUserContext(
        event(channelId: otherChannel),
        expectedAssignee: assignee,
        expectedChannelId: 'mac',
      ),
      throwsFormatException,
    );
  });

  test('a newer membership snapshot revokes an older authorised identity', () {
    final resolved = resolveAuthoritativeCosUserContextChannel(
      candidateChannelIds: const ['mac'],
      metadataEvents: [
        channelState(
          kind: 39000,
          channelId: 'mac',
          tags: const [
            ['private'],
          ],
        ),
      ],
      membershipEvents: [
        channelState(
          kind: 39002,
          channelId: 'mac',
          createdAt: 1,
          tags: [
            ['p', bridge, '', 'owner'],
            const ['p', assignee, '', 'member'],
          ],
        ),
        channelState(
          kind: 39002,
          channelId: 'mac',
          createdAt: 2,
          tags: [
            ['p', bridge, '', 'owner'],
          ],
        ),
      ],
      assigneePubkey: assignee,
      trustedBridgePubkey: bridge,
      trustedRelayPubkey: relay,
    );
    expect(resolved, isNull);
  });

  test('rejects channel authority not signed by the active relay', () {
    final resolved = resolveAuthoritativeCosUserContextChannel(
      candidateChannelIds: const ['mac'],
      metadataEvents: [
        channelState(
          kind: 39000,
          channelId: 'mac',
          tags: const [
            ['private'],
          ],
          secretKey: attackerSecret,
        ),
      ],
      membershipEvents: [
        channelState(
          kind: 39002,
          channelId: 'mac',
          tags: [
            ['p', bridge, '', 'owner'],
            const ['p', assignee, '', 'member'],
          ],
        ),
      ],
      assigneePubkey: assignee,
      trustedBridgePubkey: bridge,
      trustedRelayPubkey: relay,
    );
    expect(resolved, isNull);
  });
}
